neps/nep-0480.md

```diff
@@ -0,0 +1,143 @@
+
+- Permissions: It could be possible to set permissions on a particular namespace. For example, restricting the namespace `dao_multisig` to interactions with `dao.near` only.
+
+- Synchronous execution. Namespaced contracts on the same account should be able to communicate with each other synchronously instead of asynchronously.
+
+## Specification
+
+### The `Namespace` type
+
+We introduce a new primitive type: `Namespace`. A `Namespace` is a string that contains only lowercase letters, numbers, and underscores. The empty string is considered the default namespace.
+
+### `DeployContract` action
+
+The `DeployContract` action is modified to include a `namespace` field. If the `namespace` field is not specified in an RPC call, the default namespace (the empty string) is used. The attached code is then deployed to the specified namespace.
+
+If code has already been deployed to the specified namespace, it is replaced, leaving code deployed to other namespaces untouched.
+
+Once a namespace is created, this NEP does not provide a way to delete it. However, namespaces should not be considered permanent.
+
+### `FunctionCall` action
+
+The `FunctionCall` action is modified to include a `namespace` field. If the `namespace` field is not specified in an RPC call, the default namespace (the empty string) is used. The function call is then executed on the code deployed to the specified namespace.
+
+### `FunctionCallPermission`
+
+A `namespaces` field is added to `FunctionCallPermission`. This field contains a list of namespaces that the access key is allowed to call. If the `namespaces` field is not specified, the access key is allowed to call all namespaces.
+
+### NEAR VM host functions
+
+The following host functions are added to the NEAR host environment:
+
+- `predecessor_namespace() -> Namespace`: Returns the namespace of the predecessor account, which may be the empty namespace. In the case that the predecessor is not a smart contract, this is the empty namespace.
+- `current_namespace() -> Namespace`: Returns the namespace of the current account, which may be the empty namespace.
+- `namespace_storage_usage(&Namespace) -> u64`: Returns the storage usage of the given namespace. This includes storage consumed by the contract and state, but not account data (e.g. balance, access keys). That is to say,
+
+  $$\verb|storage_usage()| \ge \sum_{n} \verb|namespace_storage_usage| (n)$$
+
+There is no need for a `signer_namespace` function, because a signer is never a smart contract, and can therefore never have a non-empty namespace.
+
+Some existing host functions will need to be updated to include namespaces:
+
+- `promise_batch_action_function_call`: Updated to include namespace.
+- `promise_batch_action_deploy_contract`: Updated to include namespace.
+- etc.
+
+Existing host functions will generally still refer to the account as a whole:
+
+- `storage_usage() -> u64`: Returns the storage usage of the current account, including all namespaces.
```

Comment:
In the current proposal, then there are two ways of getting storage used.  `storage_usage()` and `namespace_storage_usage()`.  I can see some developers getting confused when `storage_usage()` does not add up to what they are seeing the current namespace is actually using (they might have forgotten that other namespaces exist).

In the ideal world, it would be great to have a single hostfunction with a signature like: `storage_usage(enum {EntireContract; Namespace(namespace)'})`.  Does something like this make sense?

If it does, I propose that we update the proposed `namespace_storage_usage()` to be something like the above and then mark `storage_usage()` as deprecated.  We probably never will be able to remove it but hopefully with sufficient documentation, we will minimise the confusion.

Reply:
I don't think I have enough information to have an informed opinion here. Making a decision one way or another has the potential to break contracts: if `storage_usage()` only references the current namespace, old contracts could incorrectly think the account is not using storage when it is, and if it references the whole account's usage (all namespaces), then old contracts could incorrectly think the current contract is using storage when it is not.